File: lib/fetchIndicators.ts
/**
 * Core data-fetching logic, shared between:
 *   - app/page.tsx  (server-side render, ISR-cached)
 *   - app/api/indicators/route.ts  (called by the client Refresh button)
 */
import type { ApiResponse, LiveValue } from './types';

const FRED_BASE = 'https://api.stlouisfed.org/fred/series/observations';

// Browser headers to avoid 401 from Yahoo
const BROWSER_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  'Accept': 'application/json, text/plain, */*',
  'Accept-Language': 'en-US,en;q=0.9',
  'Referer': 'https://finance.yahoo.com/',
};

// ── FRED fetcher ──────────────────────────────────────────────────────
async function fredGet(series: string, apiKey: string, limit = 20): Promise<number[]> {
  const url = `${FRED_BASE}?series_id=${series}&api_key=${apiKey}&file_type=json&sort_order=desc&limit=${limit}`;
  const res = await fetch(url, { next: { revalidate: 3600 } });
  if (!res.ok) throw new Error(`FRED ${series}: HTTP ${res.status}`);
  const data = await res.json();
  if (data.error_message) throw new Error(`FRED ${series}: ${data.error_message}`);
  return (data.observations as { value: string }[])
    .filter(o => o.value !== '.')
    .map(o => parseFloat(o.value));
}

async function fredGetLong(series: string, apiKey: string, limit = 500): Promise<number[]> {
  return fredGet(series, apiKey, limit);
}

// ── Yahoo Finance chart fetcher ───────────────────────────────────────
async function yahooChart(
  symbol: string,
  range = '2y',
  interval = '1d'
): Promise<{ timestamps: number[]; closes: number[] }> {
  const url = `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(symbol)}?interval=${interval}&range=${range}`;
  const res = await fetch(url, {
    headers: BROWSER_HEADERS,
    next: { revalidate: 3600 },
  });
  if (!res.ok) throw new Error(`Yahoo ${symbol}: HTTP ${res.status}`);
  const json = await res.json();
  const result = json?.chart?.result?.[0];
  if (!result) throw new Error(`Yahoo ${symbol}: no data`);
  const timestamps: number[] = result.timestamp || [];
  const closes: number[] = result.indicators?.quote?.[0]?.close || [];
  return { timestamps, closes };
}

// ── YoY helper (descending array: arr[0] = most recent) ──────────────
function calcYoY(arr: number[], periodsBack: number): [number, number] {
  if (arr.length < periodsBack + 2) throw new Error('Not enough data for YoY');
  const currentYoY = ((arr[0] / arr[periodsBack]) - 1) * 100;
  const priorYoY = (arr[1] && arr[periodsBack + 1])
    ? ((arr[1] / arr[periodsBack + 1]) - 1) * 100
    : currentYoY;
  return [currentYoY, priorYoY];
}

// ── S&P 500 vs 200DMA from Yahoo ──────────────────────────────────────
async function fetchSPX200DMA(): Promise<LiveValue> {
  const { closes } = await yahooChart('^GSPC', '2y', '1d'); // S&P 500
  const desc = closes.filter(v => v != null && !isNaN(v)).reverse(); // Yahoo is ascending, reverse to descending
  if (desc.length < 201) throw new Error('SPX: insufficient data');
  const ma200 = desc.slice(0, 200).reduce((s, v) => s + v, 0) / 200;
  const ma200prior = desc.slice(1, 201).reduce((s, v) => s + v, 0) / 200;
  return {
    current: parseFloat((((desc[0] / ma200) - 1) * 100).toFixed(2)),
    prior:   parseFloat((((desc[1] / ma200prior) - 1) * 100).toFixed(2)),
  };
}

// ── Copper / Gold ratio from Yahoo ────────────────────────────────────
async function fetchCopperGold(): Promise<LiveValue> {
  const [cu, au] = await Promise.all([
    yahooChart('HG=F', '2y', '1d'),  // copper futures
    yahooChart('GC=F', '2y', '1d'),  // gold futures
  ]);
  const cuClean = cu.closes.filter(v => v != null && !isNaN(v));
  const auClean = au.closes.filter(v => v != null && !isNaN(v));
  if (!cuClean.length || !auClean.length) throw new Error('Copper/Gold: missing data');
  const cuCurrent = cuClean[cuClean.length - 1];
  const auCurrent = auClean[auClean.length - 1];
  const cuPrior = cuClean.length > 1 ? cuClean[cuClean.length - 2] : cuCurrent;
  const auPrior = auClean.length > 1 ? auClean[auClean.length - 2] : auCurrent;
  return {
    current: parseFloat((cuCurrent / auCurrent).toFixed(6)),
    prior:   parseFloat((cuPrior / auPrior).toFixed(6)),
  };
}

// ── Bloomberg Commodity Index YoY from Yahoo ──────────────────────────
async function fetchBCOMYoY(): Promise<LiveValue> {
  // Try ^BCOM first, fall back to DJP (iPath Bloomberg Commodity ETN)
  let closes: number[] = [];
  try {
    const result = await yahooChart('^BCOM', '2y', '1d');
    closes = result.closes;
  } catch {
    const result = await yahooChart('DJP', '2y', '1d');
    closes = result.closes;
  }
  const desc = closes.filter(v => v != null && !isNaN(v)).reverse(); // Yahoo is ascending, reverse to descending
  if (desc.length < 253) throw new Error('BCOM: insufficient data for YoY');
  const [current, prior] = calcYoY(desc, 252);
  return {
    current: parseFloat(current.toFixed(2)),
    prior:   parseFloat(prior.toFixed(2)),
  };
}

// ── Zillow ZORI CSV → national rent YoY ──────────────────────────────
async function fetchZori(): Promise<LiveValue> {
  const urls = [
    'https://files.zillowstatic.com/research/public_csvs/zori/Metro_zori_uc_sfrcondomfr_sm_sa_month.csv',
    'https://files.zillowstatic.com/research/public_csvs/zori/Metro_zori_uc_sfrcondomfr_tier_0.33_0.67_sm_sa_month.csv',
    'https://files.zillowstatic.com/research/public_v2/zori/Metro_zori_uc_sfrcondomfr_sm_sa_month.csv',
  ];
  let text: string | null = null;
  for (const url of urls) {
    try {
      const res = await fetch(url, {
        headers: { 'User-Agent': 'Mozilla/5.0 (compatible; CIODashboard/1.0)' },
        next: { revalidate: 86400 },
      });
      if (res.ok) { text = await res.text(); break; }
    } catch { /* try next */ }
  }
  if (!text) throw new Error('Zillow CSV: all URLs failed');

  const lines = text.trim().split('\n');
  let usRow: string[] | null = null;
  for (let i = 1; i < lines.length; i++) {
    const cols = lines[i].split(',');
    if (cols[2]?.replace(/"/g, '') === 'United States') { usRow = cols; break; }
  }
  if (!usRow) throw new Error('Zillow: United States row not found');

  const header = lines[0].split(',');
  let dataStartIdx = 0;
  for (let i = 0; i < header.length; i++) {
    if (/^\d{4}-\d{2}/.test(header[i]?.replace(/"/g, ''))) { dataStartIdx = i; break; }
  }
  if (dataStartIdx === 0) dataStartIdx = 5;

  const vals: number[] = [];
  for (let i = header.length - 1; i >= dataStartIdx && vals.length < 15; i--) {
    const v = parseFloat(usRow[i]);
    if (!isNaN(v)) vals.push(v);
  }
  if (vals.length < 13) throw new Error('Zillow: insufficient data');

  const current = ((vals[0] / vals[12]) - 1) * 100;
  const prior   = (vals[1] && vals[13]) ? ((vals[1] / vals[13]) - 1) * 100 : current;
  return { current: parseFloat(current.toFixed(2)), prior: parseFloat(prior.toFixed(2)) };
}

// ── Main orchestrator ─────────────────────────────────────────────────
export async function fetchAll(apiKey: string): Promise<ApiResponse> {
  const values: Record<string, LiveValue> = {};
  const errors: Record<string, string> = {};

  async function safe<T>(key: string, fn: () => Promise<T>, transform: (v: T) => LiveValue) {
    try {
      values[key] = transform(await fn());
    } catch (e) {
      errors[key] = e instanceof Error ? e.message : String(e);
    }
  }

  await Promise.allSettled([
    // ── FRED: direct readings ──────────────────────────────────────────
    safe('GDPNOW', () => fredGet('GDPNOW', apiKey, 5), arr => ({
      current: arr[0], prior: arr[1] ?? arr[0],
    })),
    safe('WEI', () => fredGet('WEI', apiKey, 5), arr => ({
      current: arr[0], prior: arr[1] ?? arr[0],
    })),
    safe('T10Y3M', () => fredGet('T10Y3M', apiKey, 10), arr => ({
      current: parseFloat((arr[0] * 100).toFixed(1)),
      prior:   parseFloat((arr[1] * 100).toFixed(1)),
    })),
    safe('HY_SPREAD', () => fredGet('BAMLH0A0HYM2', apiKey, 10), arr => ({
      current: parseFloat((arr[0] * 100).toFixed(0)),
      prior:   parseFloat((arr[1] * 100).toFixed(0)),
    })),

    // Industrial Production Index (INDPRO)
    safe('ISM_NEW_ORDERS', () => fredGet('INDPRO', apiKey, 5), arr => ({
      current: arr[0], prior: arr[1] ?? arr[0],
    })),

    safe('ICSA',   () => fredGet('ICSA',   apiKey, 5), arr => ({ current: arr[0], prior: arr[1] ?? arr[0] })),
    safe('PERMIT', () => fredGet('PERMIT', apiKey, 5), arr => ({ current: arr[0], prior: arr[1] ?? arr[0] })),
    safe('T5YIFR', () => fredGet('T5YIFR', apiKey, 5), arr => ({ current: arr[0], prior: arr[1] ?? arr[0] })),
    safe('MICH_1YR', () => fredGet('MICH',  apiKey, 5), arr => ({ current: arr[0], prior: arr[1] ?? arr[0] })),

    // LEI — USSLIND is already an annualised 6-month growth rate (%), use directly
    safe('LEI', () => fredGet('USSLIND', apiKey, 5), arr => ({
      current: parseFloat(arr[0].toFixed(2)),
      prior:   parseFloat((arr[1] ?? arr[0]).toFixed(2)),
    })),

    // ── FRED: YoY series ──────────────────────────────────────────────
    safe('OIL_YOY', () => fredGetLong('DCOILWTICO', apiKey), arr => {
      const [c, p] = calcYoY(arr, 260);
      return { current: parseFloat(c.toFixed(1)), prior: parseFloat(p.toFixed(1)) };
    }),
    safe('PPI_YOY', () => fredGetLong('PPIACO', apiKey), arr => {
      const [c, p] = calcYoY(arr, 12);
      return { current: parseFloat(c.toFixed(2)), prior: parseFloat(p.toFixed(2)) };
    }),
    safe('DOLLAR_YOY', () => fredGetLong('DTWEXBGS', apiKey), arr => {
      const [c, p] = calcYoY(arr, 260);
      return { current: parseFloat(c.toFixed(2)), prior: parseFloat(p.toFixed(2)) };
    }),

    // ── Yahoo Finance: real-time market data ───────────────────────────
    safe('SPX_VS_200DMA', () => fetchSPX200DMA(), v => v),
    safe('COPPER_GOLD',   () => fetchCopperGold(), v => v),
    safe('BCOM_YOY',      () => fetchBCOMYoY(), v => v),

    // ── Rent YoY: Zillow ZORI → fallback to FRED rent CPI ────────────
    safe('ZORI_YOY', async () => {
      try { return await fetchZori(); }
      catch {
        const arr = await fredGetLong('CUSR0000SEHA', apiKey);
        const [c, p] = calcYoY(arr, 12);
        return { current: parseFloat(c.toFixed(2)), prior: parseFloat(p.toFixed(2)) };
      }
    }, v => v),

    // Atlanta Wage: FRED Average Hourly Earnings YoY
    safe('ATLANTA_WAGE', () => fredGetLong('AHETPI', apiKey), arr => {
      const [c, p] = calcYoY(arr, 12);
      return { current: parseFloat(c.toFixed(2)), prior: parseFloat(p.toFixed(2)) };
    }),
  ]);

  // Stamp errors so the UI can show them
  Object.keys(errors).forEach(k => {
    if (!values[k]) values[k] = { current: NaN, prior: NaN, error: errors[k] };
  });

  return { values, fetchedAt: new Date().toISOString() };
}

File: fetch_indicators.py
#!/usr/bin/env python3
"""
Fetch leading indicator data from FRED, Yahoo Finance, and Zillow.
Writes output to public/indicators.json for Vercel to serve statically.
Runs daily via GitHub Actions.

FIXES APPLIED:
- T10Y3M: Removed mul=100 (data already in percentage points)
- HY_SPREAD: Removed mul=100 (data already in percentage points)
- ISM_NEW_ORDERS: Fixed ticker INDPRO → NAPMNOI (ISM diffusion index)
- COPPER_GOLD: Returns % vs 3-month average instead of raw ratio
"""

import json
import math
import os
import time
from datetime import datetime, timezone

import requests
import yfinance as yf
from dotenv import load_dotenv

# Load .env.local for local development (GitHub Actions sets the env var directly)
load_dotenv(".env.local")

FRED_KEY       = os.environ.get("FRED_API_KEY", "").strip('"').strip("'")
FRED_BASE      = "https://api.stlouisfed.org/fred/series/observations"
FRED_META_BASE = "https://api.stlouisfed.org/fred/series"
OUT_PATH       = os.path.join(os.path.dirname(__file__), "public", "indicators.json")


# ── Helpers ────────────────────────────────────────────────────────────

def fred_get(series: str, limit: int = 20) -> dict:
    """Fetch FRED series (descending order), with up to 3 retries on 429.
    Returns observation values plus the series last_updated (publication) date."""
    for attempt in range(3):
        if attempt > 0:
            print(f"    429 on {series} — retrying in 1s...")
            time.sleep(1)
        r = requests.get(FRED_BASE, params={
            "series_id": series, "api_key": FRED_KEY,
            "file_type": "json", "sort_order": "desc", "limit": limit,
        }, timeout=30)
        if r.status_code == 429 and attempt < 2:
            continue
        r.raise_for_status()
        data = r.json()
        if "error_message" in data:
            raise ValueError(data["error_message"])
        obs = [o for o in data["observations"] if o["value"] != "."]
        obs_date = obs[0]["date"] if obs else ""

        # Fetch series metadata to get the actual publication date
        pub_date = obs_date
        try:
            meta_r = requests.get(FRED_META_BASE, params={
                "series_id": series, "api_key": FRED_KEY, "file_type": "json",
            }, timeout=30)
            if meta_r.ok:
                meta = meta_r.json()
                raw = meta.get("seriess", [{}])[0].get("last_updated", "")
                if raw:
                    pub_date = raw[:10]  # "2026-06-01 11:10:07-05" → "2026-06-01"
        except Exception:
            pass  # fall back to observation date

        return {"v": [float(o["value"]) for o in obs], "d": pub_date}
    raise RuntimeError(f"FRED {series}: rate limited after 3 attempts")


def yoy(arr: list, back: int) -> tuple:
    """YoY % change. arr[0] = most recent (descending)."""
    if len(arr) < back + 2:
        raise ValueError(f"need {back + 2} values, got {len(arr)}")
    c = (arr[0] / arr[back] - 1) * 100
    p = (arr[1] / arr[back + 1] - 1) * 100 if arr[1] and arr[back + 1] else c
    return c, p


def lv(current, prior, date: str) -> dict:
    return {"current": current, "prior": prior, "lastUpdated": date}


def safe(key: str, fn) -> dict:
    try:
        out = fn()
        print(f"  [OK] {key}")
        return out
    except Exception as e:
        print(f"  [!!] {key}: {e}")
        return {"current": None, "prior": None, "error": str(e)}


# ── Main ───────────────────────────────────────────────────────────────

def main():
    result = {}

    # ── FRED: sequential with 0.5s pauses to avoid HTTP 429 ───────────
    print("Fetching FRED series...")

    def fs(key, fn):
        result[key] = safe(key, fn)
        time.sleep(0.5)

    def direct(series, limit=5, mul=1, dec=None):
        r = fred_get(series, limit)
        v = r["v"]
        def rnd(x):
            if dec is None: return x
            if dec == 0:    return int(round(x, 0))
            return round(x, dec)
        return lv(rnd(v[0] * mul), rnd((v[1] if len(v) > 1 else v[0]) * mul), r["d"])

    def direct_yoy(series, back, dec=2):
        r = fred_get(series, 500)
        c, p = yoy(r["v"], back)
        return lv(round(c, dec), round(p, dec), r["d"])

    fs("GDPNOW",         lambda: direct("GDPNOW"))
    fs("WEI",            lambda: direct("WEI"))
    fs("T10Y3M",         lambda: direct("T10Y3M",       limit=10, mul=1, dec=2))
    fs("HY_SPREAD",      lambda: direct("BAMLH0A0HYM2", limit=10, mul=1, dec=2))
    fs("ISM_NEW_ORDERS", lambda: direct("NAPMNOI",       limit=10, dec=1))
    fs("ICSA",           lambda: direct("ICSA"))
    fs("PERMIT",         lambda: direct("PERMIT"))
    fs("T5YIFR",         lambda: direct("T5YIFR"))
    fs("MICH_1YR",       lambda: direct("MICH"))
    fs("LEI",            lambda: direct("USSLIND",       dec=2))
    fs("OIL_YOY",        lambda: direct_yoy("DCOILWTICO", 260, dec=1))
    fs("PPI_YOY",        lambda: direct_yoy("PPIACO",     12))
    fs("DOLLAR_YOY",     lambda: direct_yoy("DTWEXBGS",   260))
    fs("ATLANTA_WAGE",   lambda: direct_yoy("AHETPI",     12))

    # ── Yahoo Finance ──────────────────────────────────────────────────
    print("Fetching Yahoo Finance data...")

    def spx_200dma():
        h = yf.Ticker("^GSPC").history(period="2y")
        c = h["Close"].dropna().values
        if len(c) < 201:
            raise ValueError("insufficient data")
        return lv(
            round(((c[-1] / c[-200:].mean()) - 1) * 100, 2),
            round(((c[-2] / c[-201:-1].mean()) - 1) * 100, 2),
            h.index[-1].strftime("%Y-%m-%d"),
        )

    def copper_gold():
        cu = yf.Ticker("HG=F").history(period="2y")["Close"].dropna()
        au = yf.Ticker("GC=F").history(period="2y")["Close"].dropna()

        # Align dates (copper and gold trade on different calendars)
        merged = cu.to_frame(name="copper").join(au.to_frame(name="gold"), how="inner").dropna()
        if len(merged) < 130:
            raise ValueError("insufficient aligned data")

        merged["ratio"] = merged["copper"] / merged["gold"]
        current_ratio = merged["ratio"].iloc[-1]
        avg_3m = merged["ratio"].iloc[-63:].mean()
        pct_vs_3m = ((current_ratio - avg_3m) / avg_3m) * 100

        prior_ratio = merged["ratio"].iloc[-2]
        prior_avg_3m = merged["ratio"].iloc[-64:-1].mean()
        prior_pct_vs_3m = ((prior_ratio - prior_avg_3m) / prior_avg_3m) * 100

        return {
            "current": round(pct_vs_3m, 2),
            "prior": round(prior_pct_vs_3m, 2),
            "lastUpdated": merged.index[-1].strftime("%Y-%m-%d"),
            "raw_ratio": round(float(current_ratio), 6),
            "signal": "BULLISH" if pct_vs_3m > 0 else "BEARISH",
        }

    def bcom_yoy():
        for sym in ["^BCOM", "DJP"]:
            try:
                h = yf.Ticker(sym).history(period="2y")
                c = h["Close"].dropna().values
                if len(c) >= 254:
                    return lv(
                        round((c[-1] / c[-253] - 1) * 100, 2),
                        round((c[-2] / c[-254] - 1) * 100, 2),
                        h.index[-1].strftime("%Y-%m-%d"),
                    )
            except Exception:
                continue
        raise ValueError("insufficient data from ^BCOM and DJP")

    result["SPX_VS_200DMA"] = safe("SPX_VS_200DMA", spx_200dma)
    result["COPPER_GOLD"]   = safe("COPPER_GOLD",   copper_gold)
    result["BCOM_YOY"]      = safe("BCOM_YOY",      bcom_yoy)

    # ── Zillow ZORI (FRED rent CPI fallback) ──────────────────────────
    print("Fetching Zillow ZORI...")

    def zori():
        urls = [
            "https://files.zillowstatic.com/research/public_csvs/zori/Metro_zori_uc_sfrcondomfr_sm_sa_month.csv",
            "https://files.zillowstatic.com/research/public_csvs/zori/Metro_zori_uc_sfrcondomfr_tier_0.33_0.67_sm_sa_month.csv",
            "https://files.zillowstatic.com/research/public_v2/zori/Metro_zori_uc_sfrcondomfr_sm_sa_month.csv",
        ]
        text = None
        for url in urls:
            try:
                resp = requests.get(url, headers={"User-Agent": "Mozilla/5.0"}, timeout=30)
                if resp.ok:
                    text = resp.text
                    break
            except Exception:
                continue
        if not text:
            raise ValueError("all Zillow URLs failed")

        lines  = text.strip().split("\n")
        header = [h.strip().strip('"') for h in lines[0].split(",")]
        us_row = None
        for line in lines[1:]:
            cols = line.split(",")
            if len(cols) > 2 and cols[2].strip().strip('"') == "United States":
                us_row = [c.strip().strip('"') for c in cols]
                break
        if not us_row:
            raise ValueError("United States row not found")

        start = next((i for i, h in enumerate(header) if len(h) >= 4 and h[:4].isdigit()), 5)
        data, date = [], ""
        for i in range(len(header) - 1, start - 1, -1):
            if i >= len(us_row):
                continue
            try:
                v = float(us_row[i])
                if not math.isnan(v):
                    if not data:
                        date = header[i]
                    data.append(v)
            except ValueError:
                continue
            if len(data) >= 15:
                break
        if len(data) < 13:
            raise ValueError("insufficient data")
        return lv(
            round((data[0] / data[12] - 1) * 100, 2),
            round((data[1] / data[13] - 1) * 100, 2) if len(data) >= 14 else round((data[0] / data[12] - 1) * 100, 2),
            date,
        )

    try:
        result["ZORI_YOY"] = zori()
        print("  [OK] ZORI_YOY")
    except Exception as e:
        print(f"  [!!] Zillow: {e} -- falling back to FRED rent CPI")
        result["ZORI_YOY"] = safe("ZORI_YOY (FRED fallback)", lambda: direct_yoy("CUSR0000SEHA", 12))

    # ── Write output ───────────────────────────────────────────────────
    output = {"values": result, "fetchedAt": datetime.now(timezone.utc).isoformat()}
    os.makedirs(os.path.dirname(os.path.abspath(OUT_PATH)), exist_ok=True)
    with open(OUT_PATH, "w") as f:
        json.dump(output, f, indent=2)

    ok = sum(1 for v in result.values() if v.get("current") is not None)
    print(f"\n[DONE] {ok}/{len(result)} indicators written to {OUT_PATH}")


if __name__ == "__main__":
    main()
